### PR TITLE
docs(artifacts): raise ArtifactFinalizedError instead of ValueError

### PR DESCRIPTION
**Pull request by moredatarequired**
Fixes [WB-12545](https://wandb.atlassian.net/browse/WB-12545)

Description
-----------
Adds to #5026 and adds `ArtifactFinalizedError` when accessing methods that are invalid for finalized artifacts, and updates all the documentation accordingly.

Testing
-------
How was this PR tested?

Checklist
-------
- [x] Include reference to internal ticket "Fixes WB-NNNN" and/or GitHub issue "Fixes #NNNN" (if applicable)
- [x] Ensure PR title compliance with the [conventional commits standards](https://github.com/wandb/wandb/blob/main/CONTRIBUTING.md#conventional-commits)

[WB-12545]: https://wandb.atlassian.net/browse/WB-12545?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: wandb/sdk/wandb_run.py [hunk 1]
```diff
@@ -3669,7 +3669,7 @@ def finish(exit_code: Optional[int] = None, quiet: Optional[bool] = None) -> Non
         wandb.run.finish(exit_code=exit_code, quiet=quiet)
 
 
-class InvalidArtifact:
+class UnloggedArtifact(ArtifactInterface):
```
Comment (kptkin) [inline wandb/sdk/wandb_run.py:3672]: what is your sense of moving the artifact logic out of the wandb_run file into a dedicate artifact file?

Comment (moredatarequired) [reply wandb/sdk/wandb_run.py:3672]: I would prefer to move it elsewhere. Note my other big refactoring PR. Next move is to take the code that's here and push it out, also all the storage handlers.

Comment (moredatarequired) [reply wandb/sdk/wandb_run.py:3672]: I would prefer to move it elsewhere. Note my other big refactoring PR. Next move is to take the code that's here and push it out, also all the storage handlers.

**Review by noaleetz — APPROVED** [review]

File: wandb/sdk/interface/artifacts.py [hunk 2]
```diff
@@ -165,23 +165,50 @@ def ref_url(self) -> str:
...
+class ArtifactFinalizedError(ArtifactStatusError):
+    """Raised for Artifact methods or attributes that can't be changed after logging."""
+
+    def __init__(
+        self, artifact: Optional["Artifact"] = None, attr: Optional[str] = None
+    ):
+        super().__init__(
+            artifact,
+            attr,
+            "'{method_id}' used on logged artifact. Can't add to finalized artifact.",
```
Comment (noaleetz) [inline wandb/sdk/interface/artifacts.py:208]: Some questions from this error message:
- what is a finalized artifact? Is this a version that has already been logged and is thus immutable? (is this language we use elsewhere / does the user know what a finalized artifact is?)
- add what to a finalized artifact? Refers to adding new versions? Or adding new files to a version?
- see suggestion for rephrasing first sentence 

Sugestion: "'{method_id}' is used on an artifact that has already been logged. You cannot add ??? to this artifact as it has been finalized"

File: wandb/sdk/interface/artifacts.py [hunk 3]
```diff
@@ -165,23 +165,50 @@ def ref_url(self) -> str:
         raise NotImplementedError
 
 
-class ArtifactNotLoggedError(Exception):
-    """Raised for Artifact methods or attributes that require the logged artifact."""
+class ArtifactStatusError(AttributeError):
+    """Raised when an artifact is in an invalid state for the requested operation."""
```
Comment (noaleetz) [inline wandb/sdk/interface/artifacts.py:169]: (question) what are some invalid states that exist? Would these be artifacts that are finalized? I'm trying to understand if there should be a suggestion in there on moving the artifact to a valid state for the operation to work, or be more specific as to what kinds of invalid states might be causing the error.

**Pull request merged**